propolis/src/block.rs

```diff
@@ -217,102 +114,100 @@ impl<R: BlockReq, S: BackingStore> PlainBdev<R, S> {
-                }
-                BlockOp::Write => self.process_write_request(offset, &mem, buf),
-                BlockOp::Flush => self.process_flush(),
-            };
-
-            // If any BlockReq in this IO request fail, inform the guest that
-            // their IO request failed.
-            match result {
-                BlockResult::Success => {} // ok
-                BlockResult::Failure => {
-                    return BlockResult::Failure;
-                }
-                BlockResult::Unsupported => {
-                    return BlockResult::Unsupported;
-                }
-            };
-
-            offset += sz;
+        while let Some(buf) = req.next_buf() {
+            bufs.push(buf);
         }
 
-        BlockResult::Success
-    }
-
-    /// Delegate a block device read to BlockReqBackingStore.
-    fn process_read_request(
-        &self,
-        offset: usize,
-        sz: usize,
-        mem: &MemCtx,
-        buf: GuestRegion,
-    ) -> BlockResult {
-        if let Some(mapping) = mem.writable_region(&buf) {
-            let nread = self.backing_store.issue_read(offset, sz, mapping);
-            match nread {
-                Ok(nread) => {
-                    assert_eq!(nread as usize, sz);
-                    BlockResult::Success
-                }
-                Err(_) => {
-                    // TODO: Error writing to guest memory
-                    BlockResult::Failure
-                }
+        let result = match req.oper() {
+            BlockOp::Read => self.process_rw_request(true, offset, &mem, bufs),
+            BlockOp::Write => {
+                self.process_rw_request(false, offset, &mem, bufs)
             }
-        } else {
-            // TODO: Error getting writable region
-            BlockResult::Failure
+            BlockOp::Flush => self.process_flush(),
+        };
+
+        match result {
+            Ok(status) => status,
+            Err(_) => BlockResult::Failure,
         }
     }
 
-    /// Delegate a block device write to BlockReqBackingStore.
-    fn process_write_request(
+    /// Delegate a block device read or write to the file.
+    fn process_rw_request(
         &self,
+        is_read: bool,
         offset: usize,
         mem: &MemCtx,
-        buf: GuestRegion,
-    ) -> BlockResult {
-        let sz = buf.1;
-
-        if let Some(mapping) = mem.readable_region(&buf) {
-            let nwritten = self.backing_store.issue_write(offset, sz, mapping);
-            match nwritten {
-                Ok(nwritten) => {
-                    assert_eq!(nwritten as usize, sz);
-                    BlockResult::Success
+        bufs: Vec<GuestRegion>,
+    ) -> Result<BlockResult> {
+        let mappings: Vec<SubMapping> = bufs
+            .iter()
+            .map(|buf| {
+                if is_read {
+                    mem.writable_region(buf).unwrap()
+                } else {
+                    mem.readable_region(buf).unwrap()
```

Comment:
The unwraps aren't ideal. You could instead collect into a `Option<Vec<_>>` which would give back a None if any of the mappings were None.

Reply:
nice, done in 11152cd